### PR TITLE
Add support for OAuth client credentials

### LINKED ISSUE #198: OAuth client for provider auth
**Is your feature request related to a problem? Please describe.**

API key that is currently required for provider configuration has to be rotated manually when expired. As an alternative - OAuth was recently introduced (in beta) for all tailnets: https://github.com/tailscale/tailscale/issues/3243#issuecomment-1405360347. 
In general OAuth client provides better security experience.

**Describe the solution you'd like**
Be able to use the OAuth client for Tailscale provider auth instead of the API key.

### PR DESCRIPTION
**Pull request by clstokes**
**What this PR does / why we need it**:

Allows configuring the tailscale provider with OAuth Client ID and OAuth Client secret instead of a Tailscale API access token.

Requires changes in https://github.com/tailscale/tailscale-client-go/pull/44.

**Which issue this PR fixes**: 

Fixes #198

### CODE REVIEW

File: tailscale/provider.go [hunk 1]
```diff
@@ -68,26 +84,60 @@ func Provider(options ...ProviderOption) *schema.Provider {
...
+		client, err = tailscale.NewClient(apiKey, tailnet, tailscale.WithBaseURL(baseURL))
+		if err != nil {
+			return nil, diagnosticsError(err, "failed to initialise client")
+		}
 	}
 
 	return client, nil
 }
 
+func generateAPIKeyViaOAuth(oauthClientID string, oauthClientSecret string, baseURL string, tailnet string) (*oauth2.Token, error) {
```
[the remark this answers is not in the document]
Comment (clstokes) [reply tailscale/provider.go]: Yep, now removed, and the added `golang.org/x/oauth2` dependency.

**Review by DentonGentry — APPROVED** [review]

**Comment by clstokes** [comment]
For those following along I rebased and squashed my commits here to sign and signoff on the commits to pass the commit signature verification requirement.

**Comment by clstokes** [comment]
This needs a tagged release of `tailscale-client-go` now that https://github.com/tailscale/tailscale-client-go/pull/44 has been merged.

**Comment by clstokes** [comment]
@DentonGentry this has changed a good bit since your last review, so it's probably worthy of another look when you have the opportunity. 

Also, I don’t have permissions to hit “squash and merge” if/when you do give it a look, so will need someone to hit the button at that point.

**Review by DentonGentry — APPROVED** [review]
Were you able to test both combinations:
- apiKey not set and (oauthClientID, oauthClientSecret) set
- (oauthClientID, oauthClientSecret) not set and apiKey set

**Comment by clstokes** [comment]
Yep! 

#### API Key tests

##### Positive test

```shell
% TAILSCALE_API_KEY=masked_good_key terraform plan > /dev/null ; echo $?
0
```

##### Negative test

```shell
% TAILSCALE_API_KEY=bad_key terraform plan > /dev/null ; echo $?
╷
│ Error: Failed to fetch devices
│ 
│   with data.tailscale_devices.all,
│   on main.tf line 23, in data "tailscale_devices" "all":
│   23: data "tailscale_devices" "all" {
│ 
│ API token invalid (401)
╵
1
```

#### OAauth tests

##### Positive test

```shell
% OAUTH_CLIENT_ID=masked_client OAUTH_CLIENT_SECRET=good_secret terraform plan > /dev/null ; echo $?
0
```

##### Negative test

```shell
% OAUTH_CLIENT_ID=masked_client OAUTH_CLIENT_SECRET=bad_secret terraform plan > /dev/null ; echo $?
╷
│ Error: Failed to fetch devices
│ 
│   with data.tailscale_devices.all,
│   on main.tf line 23, in data "tailscale_devices" "all":
│   23: data "tailscale_devices" "all" {
│ 
│ Get "https://api.tailscale.com/api/v2/tailnet/-/devices": oauth2: cannot fetch token: 401 Unauthorized
│ Response: {"message":"API token invalid"}
╵
1
```

#### Combo of api key and oauth

```shell
% OAUTH_CLIENT_ID=a OAUTH_CLIENT_SECRET=b TAILSCALE_API_KEY=c terraform plan > /dev/null ; echo $?
╷
│ Error: tailscale provider credentials are conflicting - `api_key` conflicts with 'oauth_client_id' and 'oauth_client_secret'
│ 
│   with provider["github.com/tailscale/tailscale"].main,
│   on main.tf line 10, in provider "tailscale":
│   10: provider "tailscale" {
│ 
╵
1
```

**Comment by gberenice** [comment]
Fantastic! Thank you @clstokes @DentonGentry ❤️

**Comment by hatemosphere** [comment]
Waiting for a new release cut with this PR included! :yum:

**Comment by gberenice** [comment]
@DentonGentry are there any estimates when this feature is going be released? Excited to see that 🚀

**Pull request merged**